htlcswitch/link.go

```diff
@@ -2696,6 +2757,138 @@ func (l *channelLink) getDustClosure() dustClosure {
+	}
+
+	// Finally, check whether the max fee exposure was exceeded on either
+	// future commitment transaction with the fee-rate.
+	totalLocalDust := localDustSum + lnwire.NewMSatFromSatoshis(localFee)
+	if totalLocalDust > l.cfg.MaxFeeExposure {
+		return true, nil
+	}
+
+	totalRemoteDust := remoteDustSum + lnwire.NewMSatFromSatoshis(
+		remoteFee,
+	)
+	if totalRemoteDust > l.cfg.MaxFeeExposure {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// isDustWithHTLC calculates whether the proposed HTLC will make the channel
+// exceed the fee threshold. It first fetches the largest fee-rate that may be
+// on any unrevoked commitment transaction. Then, using this fee-rate,
+// determines if the to-be-added HTLC is dust. If the HTLC is dust, it adds to
+// the overall dust sum. If it is not dust, it contributes to weight, which
+// also adds to the overall dust sum by an increase in fees. If the dust sum on
+// either commitment exceeds the configured fee threshold, this function
+// returns true.
+func (l *channelLink) isDustWithHTLC(htlc *lnwire.UpdateAddHTLC,
+	incoming bool) bool {
+
+	dustClosure := l.getDustClosure()
+
+	feeRate := l.channel.WorstCaseFeeRate()
+
+	amount := htlc.Amount.ToSatoshis()
+
+	// See if this HTLC is dust on both the local and remote commitments.
+	isLocalDust := dustClosure(feeRate, incoming, true, amount)
+	isRemoteDust := dustClosure(feeRate, incoming, false, amount)
+
+	// Calculate the dust sum for the local and remote commitments.
+	localDustSum := l.getDustSum(false, fn.None[chainfee.SatPerKWeight]())
+	remoteDustSum := l.getDustSum(true, fn.None[chainfee.SatPerKWeight]())
+
+	// Grab the larger of the local and remote commitment fees.
+	commitFee := l.getCommitFee(false)
+
+	if l.getCommitFee(true) > commitFee {
+		commitFee = l.getCommitFee(true)
+	}
+
+	localDustSum += lnwire.NewMSatFromSatoshis(commitFee)
+	remoteDustSum += lnwire.NewMSatFromSatoshis(commitFee)
+
+	// Calculate the additional fee increase if this is a non-dust HTLC.
+	weight := lntypes.WeightUnit(input.HTLCWeight)
+	additional := lnwire.NewMSatFromSatoshis(
+		feeRate.FeeForWeight(weight),
+	)
+
+	if isLocalDust {
+		// If this is dust, it doesn't contribute to weight but does
+		// contribute to the overall dust sum.
+		localDustSum += lnwire.NewMSatFromSatoshis(amount)
+	} else {
+		// Account for the fee increase that comes with an increase in
+		// weight.
+		localDustSum += additional
+	}
+
+	if localDustSum > l.cfg.MaxFeeExposure {
+		// The max fee exposure was exceeded.
+		return true
+	}
+
+	if isRemoteDust {
+		// If this is dust, it doesn't contribute to weight but does
+		// contribute to the overall dust sum.
+		remoteDustSum += lnwire.NewMSatFromSatoshis(amount)
+	} else {
+		// Account for the fee increase that comes with an increase in
+		// weight.
+		remoteDustSum += additional
+	}
+
+	if remoteDustSum > l.cfg.MaxFeeExposure {
+		// The max fee exposure was exceeded.
+		return true
+	}
+
+	return false
```

Comment:
I'd recommend consolidating some of the branch execution logic in favor of something more expression oriented like this. Assigning variables once makes it easier for people to read the definition and not have to reason about operational semantics in order to understand what the code is doing.

```golang
	// localDustSum += lnwire.NewMSatFromSatoshis(commitFee) --> Delete
	// remoteDustSum += lnwire.NewMSatFromSatoshis(commitFee) --> Delete

	// Calculate the additional fee increase if this is a non-dust HTLC.
	weight := lntypes.WeightUnit(input.HTLCWeight)
	additional := lnwire.NewMSatFromSatoshis(
		feeRate.FeeForWeight(weight),
	)

        var addedLocalFees lnwire.MilliSatoshi
	if isLocalDust {
		// If this is dust, it doesn't contribute to weight but does
		// contribute to the overall dust sum.
		addedLocalFees = lnwire.NewMSatFromSatoshis(amount)
	} else {
		// Account for the fee increase that comes with an increase in
		// weight.
		addedLocalFees = additional
	}

        var addedRemoteFees lnwire.MilliSatoshi
	if isRemoteDust {
		// If this is dust, it doesn't contribute to weight but does
		// contribute to the overall dust sum.
		addedRemoteFees = lnwire.NewMSatFromSatoshis(amount)
	} else {
		// Account for the fee increase that comes with an increase in
		// weight.
		addedRemoteFees = additional
	}

        localFeeExposure := localDustSum + addedLocalFees + commitFee
        remoteFeeExposure := remoteDustSum + addedRemoteFees + commitFee

	return localFeeExposure > l.cfg.MaxFeeExposure ||
	        remoteFeeExposure > l.cfg.MaxFeeExposure
```

Reply:
don't see a problem with code as-is, it's pretty easy to understand what's going on imo

Reply:
There are two main things I'm trying to discourage:
1. `+=` and other forms of variable reassignment make it harder to reason locally about code. It also means that whatever the variable is named is wrong for some portion of the lifetime of that variable. Either it starts correct and is wrong later when reassigned, or it starts wrong and ends correct later when it is reassigned. When reassignment is present it is also not obvious which of these two scenarios we are in.
2. Early returns make it harder to grok the overall picture of what the expression is. In my offered solution we can see quite easily that if either the local or remote fee exposure blows the limit, then we are overexposed, and we can see the constituent components of the overall exposure.

This isn't a blocking comment, but I'm generally trying to reduce complexity and increase the comprehension ability of the codebase, and so I offered an alternative that solves both of these things.